### PR TITLE
DAML-LF: fix subsitution bug

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
This PR fixes substitution bug #3777 

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags, if appropriate
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Comment by remyhaemmerle-da** [comment]
@associahedron @cocreature could you check the scala I just pushed.

File: daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/TypeSubst.scala [hunk 1]
```diff
@@ -8,192 +8,55 @@ import com.digitalasset.daml.lf.language.Ast._
...
+    go(freeVars(subst), subst, typ)
+
+  private def go(fv0: Set[TypeVarName], subst0: Map[TypeVarName, Type], typ0: Type): Type =
+    typ0 match {
+      case TVar(name) => subst0.getOrElse(name, typ0)
+      case TTyCon(_) | TBuiltin(_) | TNat(_) => typ0
+      case TApp(t1, t2) => TApp(go(fv0, subst0, t1), go(fv0, subst0, t2))
+      case TForall((v0, k), t) =>
+        if (fv0.contains(v0)) {
+          val v1 = freshTypeVarName(fv0 + v0)
```
Comment (associahedron) [inline daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/TypeSubst.scala]: `fv0.contains(v0)` from above, so:

```suggestion
          val v1 = freshTypeVarName(fv0)
```

File: daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/TypeSubst.scala [hunk 2]
```diff
@@ -8,192 +8,55 @@ import com.digitalasset.daml.lf.language.Ast._
...
-      UpdateEmbedExpr(apply(typ), apply(body))
-    case UpdateLookupByKey(retrieveByKey) =>
-      UpdateLookupByKey(apply(retrieveByKey))
-    case UpdateFetchByKey(retrieveByKey) =>
-      UpdateFetchByKey(apply(retrieveByKey))
-  }
-
-  def apply(retrieveByKey: RetrieveByKey): RetrieveByKey = {
-    retrieveByKey.copy(key = apply(retrieveByKey.key))
+  def subsitute(subst: Map[TypeVarName, Type], app: TypeConApp): TypeConApp = app match {
```
Comment (associahedron) [inline daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/TypeSubst.scala]:
```suggestion
  def substitute(subst: Map[TypeVarName, Type], app: TypeConApp): TypeConApp = app match {
```

**Comment by associahedron** [comment]
Should we remove the workaround #3790 in this PR or a separate one?

**Comment by cocreature** [comment]
Separate PR is fine.

**Pull request merged**